citar-file.el

```diff
@@ -351,6 +321,21 @@ of files found in two ways:
                   nil 0 nil
                   file)))
 
+;;;; Note files
+
+(defun citar-file--get-notes-hash (&optional keys)
+  "Return hash-table with KEYS with file notes."
+  (citar-file--directory-files
+   citar-notes-paths keys citar-file-note-extensions
+   citar-file-additional-files-separator))
+
+(defun citar-file-has-notes (&optional _entries)
+  "Return predicate testing whether cite key has associated notes."
+  ;; REVIEW why this optional arg when not needed?
+  (let ((files (citar-file--get-notes-hash)))
+    (lambda (key)
+      (gethash key files))))
```

Comment:
@roshanshariff can you review this little thing on `citar-has-note`?

See my comment here.

I didn't need entries here, and I don't need them in the org-roam one.

Reply:
I'll do a more full code review later this week when I have some time, but the main reason is for commonality with the file-related functions. There it's useful to have the `entries` argument to check if an entry has the `file` field.

On the other hand, with our current notes backends, it is true that you only need the cite key. But I can imagine adding support in the future for the BibTeX [note field](https://www.bibtex.com/f/note-field/), so maybe it's a good idea to leave the API as-is and just ignore the `entries` argument if your backend doesn't need it,

Reply:
Ok, we can think on it a bit.

But if a function needs it, it can pull the entries; no?

E.g. if I understand right, doesn't seem justified to design for a hypothetical need when we have two concrete backends neither of which need it.

Reply:
Maybe this is a matter of taste to some extent. The code that's calling `has-notes` has a set of entries that it is operating on. To me, it seems cleaner to pass that on to the backend, which can either use it or ignore it. The alternative is for the backend to call back into Citar to get the entries, but this is not trivial. It involves asking the major-mode function for the list of bibliographies, getting them from the cache, merging their entries into another big hash table (which can use a lot of memory for no reason), etc. This is the reason some other API functions accept an optional `:entries` keyword argument, so they can avoid calling `citar-get-entries` if the `entries` hash table is already available.

And finally, it's easier to unit test a function just by passing it a hash table, knowing that its behaviour only depends on the arguments you pass in (rather than having to set up an environment where `citar-get-entries` returns the right thing).

In general, I prefer functions that take arguments for the information they need and return results that depend only on those arguments rather than global state. Of course, global state is unavoidable, but I think it's better for testability, reasoning, etc. if access to it is not scattered throughout the code base but carefully controlled to a few places.

Reply:
Ok. In any case, I'll defer to you; just thought I'd raise it.

Reply:
Sure, and of course I might come around to a different view later. But also see my edit above about race conditions.

In general, accessing the cache is not "free", especially now that it can unpredictably decide to re-parse files. In general, the correct way to use it is to get everything you need from the cache at the beginning of some operation (i.e. call `citar-get-entries` in this case), use that information throughout, and then throw it away only when you're done.

Reply:
In other news, I fixed at least one obvious bug (that was causing a compilation error noted in the MELPA review) in the capf. Perhaps it works now?

I can't easily test it.

Reply:
> I'll do a more full code review later this week when I have some time ...

I'm not sure if it's useful, but I just added [this branch](https://github.com/emacs-citar/citar/tree/pre-1.0-cleanup).

You should have write access, if you need/want it.